pkg/batchfetcher/batchfetcher.go

```diff
@@ -0,0 +1,131 @@
+package batchfetcher
+
+import (
+	"fmt"
+
+	availabilitydsl "github.com/filecoin-project/mir/pkg/availability/dsl"
+	availabilityevents "github.com/filecoin-project/mir/pkg/availability/events"
+	"github.com/filecoin-project/mir/pkg/dsl"
+	"github.com/filecoin-project/mir/pkg/events"
+	"github.com/filecoin-project/mir/pkg/modules"
+	"github.com/filecoin-project/mir/pkg/pb/availabilitypb"
+	"github.com/filecoin-project/mir/pkg/pb/eventpb"
+	"github.com/filecoin-project/mir/pkg/pb/requestpb"
+	t "github.com/filecoin-project/mir/pkg/types"
+)
+
+// NewModule returns a new batch fetcher module.
+// The batch receives events output by the ordering protocol (e.g. ISS)
+// and relays them to the application in the same order.
+// It replaces the DeliverCert events from the input stream by the corresponding ProvideTransactions
+// that it obtains from the availability layer.
+// It keeps track of the current epoch (by observing the relayed NewEpoch events)
+// and automatically requests the transactions from the correct instance of the availability module.
+func NewModule(mc *ModuleConfig) modules.Module {
+	m := dsl.NewModule(mc.Self)
+
+	// The current epoch number, as announced by the ordering protocol.
+	epochNr := t.EpochNr(0)
+
+	// Map of delivered requests that is used to filter duplicates.
+	// TODO: Implement compaction (client watermarks) so that this map does not grow indefinitely.
+	delivered := make(map[t.ClientID]map[t.ReqNo]struct{})
+
+	// Queue of output events. It is required for buffering events being relayed
+	// in case a DeliverCert event that has arrived before has not yet been transformed to a ProvideTransactions event.
+	var output outputQueue
+
+	// The NewEpoch handler updates the current epoch number and forwards the event to the output.
+	dsl.UponEvent[*eventpb.Event_NewEpoch](m, func(newEpoch *eventpb.NewEpoch) error {
+		epochNr = t.EpochNr(newEpoch.EpochNr)
+		output.Enqueue(&outputItem{
+			event: events.NewEpoch(mc.Destination, t.EpochNr(newEpoch.EpochNr)),
+		})
+		output.Flush(m)
+		return nil
+	})
+
+	// The DeliverCert handler requests the transactions referenced by the received availability certificate
+	// from the availability layer.
+	// TODO: Verify Certificate? Here or elsewhere?
```

Comment:
I think it is ok to assume that the certificate is valid here (i.e., to push the responsibility of verifying them to the module that sends the `DeliverCert` events).

Reply:
Agree. Adapted comment.